libraries/AP_Proximity/AP_Proximity_RPLidarA2.cpp

```diff
@@ -165,114 +159,35 @@ void AP_Proximity_RPLidarA2::send_request_for_device_info()
-    _byte_count = 0;
-    return false;
-}
-
 void AP_Proximity_RPLidarA2::get_readings()
 {
-    Debug(2, "             CURRENT STATE: %u ", (unsigned)_state);
-    const uint32_t nbytes = _uart->available();
-    if (nbytes == 0) {
-        return;
-    }
-    const uint32_t bytes_to_read = MIN(nbytes, sizeof(_payload)-_byte_count);
-    if (bytes_to_read == 0) {
-        INTERNAL_ERROR(AP_InternalError::error_t::flow_of_control);
-        reset();
-        return;
-    }
-    const uint32_t bytes_read = _uart->read(&_payload[_byte_count], bytes_to_read);
-    if (bytes_read == 0) {
-        // this is bad; we were told there were bytes available
-        INTERNAL_ERROR(AP_InternalError::error_t::flow_of_control);
-        reset();
-        return;
-    }
-    _byte_count += bytes_read;
-
-    uint32_t previous_loop_byte_count = UINT32_MAX;
-    while (_byte_count) {
-        if (_byte_count >= previous_loop_byte_count) {
-            // this is a serious error, we should always consume some
-            // bytes.  Avoid looping forever.
-            INTERNAL_ERROR(AP_InternalError::error_t::flow_of_control);
-            _uart = nullptr;
-            return;
-        }
-        previous_loop_byte_count = _byte_count;
-
+    Debug(2, "CURRENT STATE: %u ", (unsigned)_state);
+    while (_uart->available()) {
         switch(_state){
         case State::RESET: {
-            // looking for 0x52 at start of buffer; the 62 following
-            // bytes are "information"
-            if (!make_first_byte_in_payload('R')) { // that's 'R' as in RPiLidar
-                return;
-            }
-            if (_byte_count < 63) {
+            if (AP_HAL::millis() - _last_reset_ms < 1000) {
                 return;
             }
-#if RP_DEBUG_LEVEL
-            // optionally spit out via mavlink the 63-bytes of cruft
-            // that is spat out on device reset
-            Debug(1, "Got RPLidar Information");
-            char xbuffer[64]{};
-            memcpy((void*)xbuffer, (void*)&_payload.information, 63);
-            gcs().send_text(MAV_SEVERITY_INFO, "RPLidar: (%s)", xbuffer);
-#endif
-            // 63 is the magic number of bytes in the spewed-out
-            // reset data ... so now we'll just drop that stuff on
-            // the floor.
-            consume_bytes(63);
+            _uart->discard_input();
             send_request_for_device_info();
             _state = State::AWAITING_RESPONSE;
             continue;
         }
         case State::AWAITING_RESPONSE:
-            if (_payload[0] != RPLIDAR_PREAMBLE) {
-                // this is a protocol error.  Reset.
-                reset();
+            // descriptor packet has 7 byte in total
+            if (_uart->available() < sizeof(_descriptor)) {
                 return;
             }
 
-            // descriptor packet has 7 byte in total
-            if (_byte_count < sizeof(_descriptor)) {
+            _uart->read(&_payload[0], sizeof(_descriptor));
```

Comment:
I think maybe we should change this to read (or peek) just a single byte until we find the RPLIDAR_PREAMBLE byte.  As it is, it always reads 7 bytes at a time so if a stray byte gets in there it will only recover if it gets lucky and the preamble byte of some later packet happens to be at a 7-byte interval in the buffer.